src/compiler/checker.ts

```diff
@@ -15666,11 +15674,22 @@ namespace ts {
             return !!(getGenericObjectFlags(type) & ObjectFlags.IsGenericIndexType);
         }
 
+        function isUnionWithGenericDiscriminantProperty(type: UnionType) {
+            getPropertiesOfType(type); // initialize property cache (includes partial props)
+            const props = arrayFrom((type.propertyCache || emptySymbols).values());
+            for (const prop of props) {

```

Comment:
Is the for loop better than 

```ts
return props.find(prop => isDiscriminantProperty(type, prop.escapeName) === DiscriminantKind.Potential)
```

?

Reply:
Ehh... It probably can be if I swap the loop to using the iterator directly rather than collecting it, like it was thinking.

Reply:
I thought we were switching to target ES6 sometime soon? At least, I'm pretty sure that would let us use `for-of` to iterate maps finally.